engine/lib/luajit-ffi-gen/README.md

```diff
@@ -150,11 +165,92 @@ For the variants without values starting index can be set, otherwise it starts f
 
 If `repr` parameter is set then `#[repr(...)]` attribute will be added with the specified type, otherwise type will be deducted from the maximal variant value: u8, u16, u32 or u64.
 
+## Joining enum and impl blocks
+
+If it's required to expose both `enum` and its `impl` block then `with_impl` attribute should be used:
+
+```rust
+#[luajit_ffi_gen::luajit_ffi(with_impl = true)]
+#[derive(Debug)]
+pub enum MyEnum1 {
+    Var1,
+    Var2,
+}
+
+#[luajit_ffi_gen::luajit_ffi]
+impl MyEnum1 {
+    pub fn is_var1(&self) -> bool {
+        *self == Self::Var1
+    }
+}
+```
+
+This will generate following C API wrappers:
+```rust
+#[no_mangle]
+pub const MyEnum1_Var1: u8 = MyEnum1::Var1.value();
+
+#[no_mangle]
+pub const MyEnum1_Var2: u8 = MyEnum1::Var2.value();
+
+#[no_mangle]
+pub extern "C" fn MyEnum1_ToString(this: MyEnum1) -> *const libc::c_char {
+    // ...
+}
+
+#[no_mangle]
+pub extern "C" fn MyEnum1_IsVar1(this: &MyEnum1) -> bool {
+    this.is_var1()
+}
+```
+
+and **MyEnum1.lua**:
+```lua
+-- My_Enum1 --------------------------------------------------------------------
+local ffi = require('ffi')
+local libphx = require('ffi.libphx').lib
+local MyEnum1
+
+do -- C Definitions
+    ffi.cdef [[
+        typedef uint8 MyEnum1;
+
+        MyEnum1 MyEnum1_Var1;
+        MyEnum1 MyEnum1_Var2;
+
+        cstr     MyEnum1_ToString(MyEnum1);
+
+        bool MyEnum1_IsVar1(MyEnum1);
+    ]]
+end
+
+do -- Global Symbol Table
+    MyEnum1 = {
+        Var1     = libphx.MyEnum1_Var1,
+        Var2     = libphx.MyEnum1_Var2,
+
+        ToString = libphx.MyEnum1_ToString,
+
+        IsVar1 = libphx.MyEnum1_IsVar1,
+    }
+
+    if onDef_MyEnum1 then onDef_MyEnum1(MyEnum1, mt) end
+    MyEnum1 = setmetatable(MyEnum1, mt)
+end
+
+return MyEnum1
+```
+
+Take in account that `enum` block should be defined before `impl` otherwise `enum` data will be lost. Also `opaque` parameter doesn't have any influence in this case.
+
+Under the hood proc macro on the `enum` block instead of generating ***.lua** script saves all necessary information in JSON file in **target/ffi** folder. Thi information is merged later by proc macro on `impl` block.
```

Comment:
```suggestion
Under the hood proc macro on the `enum` block instead of generating ***.lua** script saves all necessary information in JSON file in **target/ffi** folder. This information is merged later by proc macro on `impl` block.
```